### ISSUE TITLE
Add lazy variant of JavaVersion.toVersion()

### ISSUE THREAD

**Issue by lptr** [gradle/gradle#25671]
Currently `JavaVersion.toVersion(Object)` does not support resolving `Provider`s. We should have a variant of this method that takes a `Provider<?>` and resolves it to a `Provider<JavaVersion>`.

Once we introduce this new method we should check where `toVersion(Object)` can be replaced with it (e.g. in `ScalaBasePlugin`).

Note that `JavaVersion` currently lives in `:base-services` that has no access to `Provider` defined in `:core-api`. We could move `JavaVersion` to `:core-api` as well (where its package `org.gradle.api` would suggest it should live). However, some functionality of `JavaVersion` is used in `:base-services` and other places that don't have access to `:core-api`. We could move this part of the logic out from `JavaVersion`; it's mostly about checking if the current Java version is compatible with Java 7 or 9 etc. The remaining part of `JavaVersion` could then be moved to `:core-api`.

---
cc: @gradle/bt-execution

**Comment by lptr** [comment]
Do we need to make `JavaVersion.toVersion()` "provider-aware" or can we use `Provider<String>.map(JavaVersion.toVersion)` or something similar instead?

**Comment by asodja** [comment]
More info: 
This is source compatibility issue.

Main problem for a user is that `JavaVersion.toVersion(Object value)` accepts `Object` instead of some exact type.
And when we will migrate for example `sourceCompatiblity` to a `Property<String>` ,  this method will call `toString()` on the provider and will fail. These failures could be discovered only at runtime.

If we want to be 100% source compatible, we should call `.get()` on a `Provider` instead of `.toString()`.

I a good solution could also be, to change  `JavaVersion.toVersion(Object value)`  to accept `JavaVersion.toVersion(String value)`  and  `JavaVersion.toVersion(int value)`  and some other supported cases.
That way users will get compile errors if they pass incorrect type.  And they when they use it with providers they could use: `Provider<String>.map(JavaVersion.toVersion)`.

**Comment by lptr** [comment]
One option is to let `JavaVersion.toVersion(Object)` resolve `Provider`s, but show a deprecation warning. But personally I think we should just resolve `Provider`s and be done with it.

Meanwhile, we could think about introducing strongly types overloads for `String` and `int`.

**Comment by octylFractal** [comment]
I agree with allowing it but with a deprecation warning. I don't think it's a good idea to keep it around as it's a non-obvious termination of the laziness of a `Provider`.

**Comment by lptr** [comment]
This is blocking upgrade of `sourceCompatibility`; if we don't do it via PAPI migration, this might not be needed for the migration itself.
